### PR TITLE
FEATURE: Auto-remove users without permission from channel

### PR DESCRIPTION
**Pull request by martin-brennan**
There are many situations that may cause users to lose permission to
send messages in a chat channel. Until now we have relied on security
checks in `Chat::ChatChannelFetcher` to remove channels which the
user may have a `UserChatChannelMembership` record for but which
they do not have access to.

This commit takes a more proactive approach. Now any of these following
`DiscourseEvent` triggers may cause `UserChatChannelMembership`
records to be deleted:

* `category_updated` - Permissions of the category changed
   (i.e. CategoryGroup records changed)
* `user_removed_from_group` - Means the user may not be able to access the
   channel based on `GroupUser` or also `chat_allowed_groups`
* `site_setting_changed` - The `chat_allowed_groups` was updated, some
   users may no longer be in groups that can access chat.
* `group_destroyed` - Means the user may not be able to access the
   channel based on `GroupUser` or also `chat_allowed_groups`

All of these are handled in a distinct service run in a background
job. Users removed are logged via `StaffActionLog` and then we
publish messages on a per-channel basis to users who had their
memberships deleted.

When the user has a channel they are kicked from open, we show
this dialog.

When they click OK we redirect them either:

* To their first other public channel, if they have any followed
* The chat browse page if they don't

This is to save on tons of requests from kicked out users getting messages
from other channels.

When the user does not have the kicked channel open, we can just
silently yoink it out of their sidebar and turn off subscriptions.

### CODE REVIEW

**Comment by jjaffeux** [comment]
@martin-brennan we had a call with @Flink about your questions which are mostly:

- how do we handle calling a service from a service
- how do we handle calling a service from a job

I will answer the first here, and @Flink will answer the second one, and also explains why he thinks we actually don't need the top level service here.

---

So, how do we handle calling a service from a service? short answer: we don't. Conceptually, services are an orchestrator of logic, they shouldn't have too much logic themselves.

- model fetching
- input handling
- validation
- access
- send_notification
- update_channel
- ...

The important part here is `send_notification`, `update_channel`... What if this `send_notification` is actually what a service I have is exactly doing? Of course, the initial reaction is: I certainly don't want to duplicate this logic, I will just call my service.

Most likely, the core business logic of `SendNotification` should be extracted as a small dedicated action, in some library or model (we might introduce a concept for this at some point). The important part of services is that you have a known state: you know your model, your data and have done access checks. So you can have very dedicated actions which only do what their name say.

Generally speaking, if your step is huge, it should be a hint that you are trying to put too much logic into your service and some of it should be better extracted into smaller objects like actions or even query objects.

I could accept one top level meta service objects having each step being a service in rare cases, but the path to every service objects can call other services objects and so on, is a path I would consider dangerous and caused by incorrect abstractions.

File: plugins/chat/app/services/auto_remove_membership_event_handler.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove_membership_event_handler.rb:21]: Actually, I don’t think we need this service at all :sweat_smile: 
Since we’re the ones providing the events with their data in `plugin.rb`, we could just call the right service (or job calling the right service) instead. We’ll end up with three jobs instead of one, but this would remove a level of indirection. And it removes the need to call a service from a service :wink:

**Comment by Flink** [comment]
So, on “how do we handle calling a service from a job”, I think there are two possibilities:
- the first one is we just call the service, it’s a “fire and forget” one, so we don’t really care about its result.
- second one is to adapt our `with_service` helper to make it work with jobs. I think it should be relatively easy to do (I’m going to conduct some tests). This way we could simply match on some failures we’re interested in and decide what to do. It could be retrying the job, raising an exception, logging something, etc.

**Comment by martin-brennan** [comment]
@jjaffeux and @Flink I do see your reasoning on not needing a service for `AutoRemoveMembershipEventHandler`, however I don't think we will be avoid services calling other services forever, which is why I was wondering about what we do in that scenario. I don't think it will be as cut and dry as "no services calling services".

If we do stay on that line of thinking, we need to make it super clear to engineers what _should_ and what _should not_ be a service, i.e. where is the line between making one and putting things into more of a plain ruby library class, and have clear guidance around why service -> service calls are bad, because people will ask for sure, especially at the point this infrastructure moves into core.

So, I will remove `AutoRemoveMembershipEventHandler` in this case, but we need to have a good answer about the details above.

@Flink and yes adapting the `with_service` helper will be really helpful, especially in this case. We can determine whether to retry etc. based on the failure. Like I said, if the `user_id` is missing or the user is deleted for `HandleUserRemovedFromGroup`, there is nothing we can do, but maybe there is some other reason for failure that we can retry on. Let me know when that is ready, in the meantime I will continue refining this PR and adding tests.

Thanks both of you for thinking about this :+1:

File: plugins/chat/app/services/auto_remove_membership_event_handler.rb
Comment (martin-brennan) [reply plugins/chat/app/services/auto_remove_membership_event_handler.rb:21]: @Flink this is not used now, I am just keeping it for my own reference for the time being with its spec, but will delete it before I send this for proper review.

File: app/models/group.rb
Comment (martin-brennan) [inline app/models/group.rb:916]: Without this I would need to check permissions of every user in chat if a group is destroyed, since all the dependent destroys remove any clues

File: plugins/chat/app/jobs/regular/auto_join_channel_memberships.rb
Comment (martin-brennan) [inline plugins/chat/app/jobs/regular/auto_join_channel_memberships.rb:5]: Renamed this since `Manage` makes it sound like it adds _and_ removes and I didn't want any confusion

File: plugins/chat/assets/javascripts/discourse/services/chat-subscriptions-manager.js
Comment (martin-brennan) [inline plugins/chat/assets/javascripts/discourse/services/chat-subscriptions-manager.js:144]: Yes I realise this is not ideal how this is interacting with the UI -- I can refactor this once we extract the message actions and "panels" out as per https://github.com/discourse/discourse/pull/20324

File: plugins/chat/app/services/auto_remove/handle_chat_allowed_groups_change.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_chat_allowed_groups_change.rb:72]: Same comment as earlier about using `Hash.new {|h, k| h[k] = [] }`.

File: plugins/chat/app/services/auto_remove/handle_chat_allowed_groups_change.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_chat_allowed_groups_change.rb:77]:
```suggestion
          context[:users_removed_map] = users_removed_map
```

File: plugins/chat/app/services/auto_remove/handle_destroyed_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_destroyed_group.rb:32]: Here also we should probably add some validation, otherwise this contract isn’t useful.

File: plugins/chat/app/services/auto_remove/handle_destroyed_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_destroyed_group.rb:51]: I definitely think this should be extracted as a policy :slightly_smiling_face:

File: plugins/chat/app/services/auto_remove/handle_destroyed_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_destroyed_group.rb:81]: Same comment as earlier about `Hash.new {|h, k| h[k] = [] }`.

File: plugins/chat/app/services/auto_remove/handle_destroyed_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_destroyed_group.rb:86]:
```suggestion
          context[:users_removed_map] = users_removed_map
```

File: plugins/chat/app/services/auto_remove/handle_destroyed_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_destroyed_group.rb:99]: Same comment as earlier about `Hash.new {|h, k| h[k] = [] }`.

File: plugins/chat/app/services/auto_remove/handle_destroyed_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_destroyed_group.rb:104]:
```suggestion
          context[:users_removed_map] = users_removed_map
```

File: plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb:31]: We should probably rewrite this contract as:
```ruby
attribute :user_id, :integer

validates :user_id, presence: true
```

File: plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb:41]: This one could be extracted as a policy.

File: plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb:59]:
```suggestion
            context[:users_removed_map] = users_removed_map
```

File: plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb
Comment (Flink) [inline plugins/chat/app/services/auto_remove/handle_user_removed_from_group.rb:80]:
```suggestion
          context[:users_removed_map] = users_removed_map
```

File: plugins/chat/plugin.rb
Comment (Flink) [inline plugins/chat/plugin.rb:628]: Here we’re already sure about having a `category_channel` so we could just do `if category_channel.auto_join_users?`

File: plugins/chat/plugin.rb
Comment (Flink) [inline plugins/chat/plugin.rb:627]: We could do an early return here, wdyt?

**Review by Flink — COMMENTED** [review]
Here is a first batch of comments, mostly small suggestions/improvements (lots of duplicate comments). I haven’t looked at specs yet.

But there’s a pattern we’re repeating a lot in those services which is determine which users to remove, destroy them, make a hash then publish events. We already have two actions but I think there’s room for improvement/refactor. I’ll play with the code and try to see if I can make some “real” improvements (I’ll keep you posted anyway even if I’m not able to manage to do anything meaningful :sweat_smile:).

**Comment by Flink** [comment]
Ok, I’ve played with the code a little and better understood some parts of it. I think some of my comments can be discarded. Instead I’ve pushed a draft PR against yours @martin-brennan to show you some refactoring suggestions :grin: 
https://github.com/discourse/discourse/pull/20484

I’ve only addressed the simplest services but I’m pretty sure we can use a similar approach to the other ones.
Tell me what you think :slightly_smiling_face: (I could work on this tomorrow if needed or I’ll leave you to it if you prefer)

File: plugins/chat/app/serializers/chat_channel_serializer.rb
Comment (martin-brennan) [reply plugins/chat/app/serializers/chat_channel_serializer.rb:120]: Ah you're right I missed that, tnaks

File: plugins/chat/app/services/actions/calculate_memberships_for_removal.rb
Comment (martin-brennan) [reply plugins/chat/app/services/actions/calculate_memberships_for_removal.rb:81]: Yes that's better, I had done that before with the settings `_map` method and forgot, thanks

File: plugins/chat/app/services/auto_remove/handle_category_updated.rb
Comment (martin-brennan) [reply plugins/chat/app/services/auto_remove/handle_category_updated.rb:39]: Oh yes that's much better than the `noop` spam, thanks

File: plugins/chat/app/services/actions/auto_removed_user_publisher.rb
Comment (martin-brennan) [reply plugins/chat/app/services/actions/auto_removed_user_publisher.rb:11]: Yep sounds good :+1:

File: plugins/chat/app/services/auto_remove/handle_category_updated.rb
Comment (martin-brennan) [reply plugins/chat/app/services/auto_remove/handle_category_updated.rb:68]: Ah okay I thought we always had to use merge, thanks

**Comment by martin-brennan** [comment]
> I’ve only addressed the simplest services but I’m pretty sure we can use a similar approach to the other ones.
Tell me what you think slightly_smiling_face (I could work on this tomorrow if needed or I’ll leave you to it if you prefer)

Thanks @Flink for the review and suggestions I agree with everything you've added here, will take what you've got in that PR and run with it and apply it (and the other changes) here. Just leave it to me :)

**Comment by martin-brennan** [comment]
One thing -- this does not work as a `model`, since it can sometimes return `[]` which is perfectly valid, we don't want to fail in that case. So it just needs to be a step that adds to context:

image

**Comment by martin-brennan** [comment]
Also not sure I 100% agree with removing the contract here:

image

Even if the contract doesn't do much validation I think it's still useful to have around to indicate what params the class needs?

File: app/models/group.rb
Comment (martin-brennan) [reply app/models/group.rb:921]: I dislike these in general too because it makes searching harder

File: plugins/chat/plugin.rb
Comment (martin-brennan) [reply plugins/chat/plugin.rb:628]: Oh yeah not sure why I kept that, thanks

File: plugins/chat/plugin.rb
Comment (martin-brennan) [reply plugins/chat/plugin.rb:627]: Yep sure :+1:

File: plugins/chat/app/services/auto_remove/handle_category_updated.rb
Comment (martin-brennan) [reply plugins/chat/app/services/auto_remove/handle_category_updated.rb]: Though I did have to use `assign_defaults` in certain cases...like `HandleDestroyedGroup`, there are two steps. If I extract some of the AR finding models/querying into a `model` step, if the model fails to return anything (which is fine on the first step) then the second step will never be run, but I need the second step to be run too.

File: plugins/chat/plugin.rb
Comment (martin-brennan) [inline plugins/chat/plugin.rb:625]: Had to keep this as a normal `if` statement, early returns caused this error to occur in specs:

```
 25) Chat::Api::ChatChannelThreadsController show when thread exists when user cannot access the channel returns 403
      Failure/Error: DiscourseEvent.on(event_name) { |*args, **kwargs| block.call(*args, **kwargs) if enabled? }

      LocalJumpError:
        unexpected return
      # ./plugins/chat/plugin.rb:627:in `block (2 levels) in activate!'
      # ./lib/plugin/instance.rb:491:in `block in on'
      # ./lib/discourse_event.rb:12:in `block in trigger'
      # ./lib/discourse_event.rb:12:in `trigger'
      # ./spec/support/discourse_event_helper.rb:5:in `trigger'
      # ./app/models/category.rb:947:in `block (2 levels) in <class:Category>'
```

Not sure how the original `return if !category.is_a?(Category)` worked, maybe there was no spec for it or it just never happened.

**Comment by martin-brennan** [comment]
@Flink ready for another review here please :)

File: plugins/chat/app/services/auto_remove/handle_category_updated.rb
Comment (Flink) [reply plugins/chat/app/services/auto_remove/handle_category_updated.rb]: Indeed, those services are a bit trickier.

**Comment by Flink** [comment]
> One thing -- this does not work as a `model`, since it can sometimes return `[]` which is perfectly valid, we don't want to fail in that case. So it just needs to be a step that adds to context:
> 
> image

Returning `[]` may be valid but then if that’s the case, the service doesn’t have any work to do :slightly_smiling_face: We can inspect quite easily the result object, knowing that it failed at the `memberships_to_remove` step, meaning there were nothing to remove. And we can simply choose to ignore this failure.

> Also not sure I 100% agree with removing the contract here:
> 
> image
> 
> Even if the contract doesn't do much validation I think it's still useful to have around to indicate what params the class needs?

I ’d say this is the documentation responsibility more than the contract one. If there is no need for coercion or for validation, then a contract isn’t useful at all. Here at first I thought we could put a presence validation but when I ran the specs I saw that an empty string was actually a valid value. So no need for a contract in the end :sweat_smile:

File: plugins/chat/plugin.rb
Comment (Flink) [reply plugins/chat/plugin.rb:625]: Ah yes, that’s because we’re in a block context. Just replace `return` by `break` and it should work as expected :slightly_smiling_face:

**Comment by Flink** [comment]
@martin-brennan I’ve just updated https://github.com/discourse/discourse/pull/20484 as I’ve found a way to refactor the more complex services, I think you’ll like it :grin: 
I’m starting to look at specs :sweat_smile:

**Comment by Flink** [comment]
@martin-brennan I’ve added a refactor for one of the specs in https://github.com/discourse/discourse/pull/20484/files#diff-2541a9289f9ea57861cfeaa624e66050ec5ffa4e6a511d714b12e41a11d05edb, tell me what you think :slightly_smiling_face:  I’ve mainly restructured things a bit and used some dedicated matchers (`have_attributes`, `expect { … }.to change { … }`) to simplify things a little.

**Comment by martin-brennan** [comment]
> Returning [] may be valid but then if that’s the case, the service doesn’t have any work to do slightly_smiling_face We can inspect quite easily the result object, knowing that it failed at the memberships_to_remove step, meaning there were nothing to remove. And we can simply choose to ignore this failure.

@Flink I should have been clearer -- this is for one of the handlers that has multiple steps, e.g. `HandleDestroyedGroup#remove_users_outside_allowed_groups`. It works fine for `HandleChatAllowedGroupsChange` though.

>  I’ve mainly restructured things a bit and used some dedicated matchers (have_attributes, expect { … }.to change { … }) to simplify things a little.

Nice, I didn't know about `have_attributes`. Will apply these concepts to all the specs for the handlers.

> I’ve found a way to refactor the more complex services

TBH I'm sorry but I prefer just having 2 steps rather than everything in a big chain in one step like https://github.com/discourse/discourse/pull/20484/files#diff-82f805b8f2423fbb60a23e451ec0021f541112130d07c874b8a601a79e407420  . Easier to reason about "it removes users because of this, and then it removes users because of this"

**Comment by jjaffeux** [comment]
Yes, don't want to keep this around for too long, let's settle on something you are comfortable supporting @martin-brennan, I will review this extensively tomorrow.

Thanks for the all inputs @Flink 🙌

**Comment by martin-brennan** [comment]
Just pushed a last commit to conform the specs to recommendations in https://github.com/discourse/discourse/pull/20484

**Comment by Flink** [comment]
> this is for one of the handlers that has multiple steps

Yes, indeed but as the change hasn’t been applied in `HandleCategoryUpdated` I thought you were referring to that one actually :sweat_smile: So this means this is probably an oversight?

> Easier to reason about "it removes users because of this, and then it removes users because of this"

Well, yes, maybe this is something that could be split to make it more explicit. But the thing I wanted to show is that we don’t need to have a default assignment and use early returns with steps that are doing more or less the same thing. The decision logic about which memberships to remove can be made in one place, and we can leverage `model` so we don’t have to handle ourselves what happens when there are no memberships to remove. We could imagine having two steps to fetch the memberships (one using `Actions::CalculateMembershipsForRemoval` and another one using AR query methods) and use explicit naming for them. Then maybe we could use `model` to merge results, this way if the final membership collection is still empty, the service will stop here. I’m thinking out loud here, just to be clear :slightly_smiling_face:

File: plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb:20]: Leftover :wink:

File: plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb:41]: We have two `before` blocks defined here. We should move this line into the `before` block line 43.

File: plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb:116]: Same remark as before, we should merge those `before` blocks together.

File: plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/auto_remove/handle_destroyed_group_spec.rb:138]: Same remark about `before` blocks :sweat_smile:

File: plugins/chat/spec/services/auto_remove/handle_user_removed_from_group_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/auto_remove/handle_user_removed_from_group_spec.rb:19]: Leftover :wink:

File: plugins/chat/spec/services/auto_remove/handle_user_removed_from_group_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/auto_remove/handle_user_removed_from_group_spec.rb:235]: We could just write `it { is_expected.to fail_a_policy(:user_not_staff) }` as it’s relatively explicit and the description is very similar to `it fails the user_not_staff policy`. Maybe that’s on purpose?

**Review by Flink — COMMENTED** [review]
Just some small comments, nothing too serious :slightly_smiling_face:

File: plugins/chat/spec/services/auto_remove/handle_user_removed_from_group_spec.rb
Comment (martin-brennan) [reply plugins/chat/spec/services/auto_remove/handle_user_removed_from_group_spec.rb:235]: Yes sorry I forget about the `it {is_expected}` format most of the time, will change this and other instances :)

**Comment by martin-brennan** [comment]
@Flink I addressed the comments again now :)

**Review by Flink — APPROVED** [review]

**Comment by martin-brennan** [comment]
@jjaffeux merged master into this and fixed up module loading issues for zeitwork, not sure if you want to double check, if so feel free :)

**Comment by martin-brennan** [comment]
Backend test failure is unrelated

**Pull request merged**
